run-network/creator.py

```diff
@@ -0,0 +1,461 @@
+            proc_str = re.sub(
+                r"    volumes:",
+                f'    volumes:\n      - "{vol}"',
+                proc_str,
+            )
+
+        for env in p.environment:
+            proc_str = re.sub(
+                r"    environment:",
+                f"    environment:\n      - {env}",
+                proc_str,
+            )
+
+        res += proc_str
+    return res
+
+
+def create_env(env: Type[ForwardRef("Environment")], dest: str):
+    """Create .env file to configure docker-compose
+
+    Info is read from the config-object and written to the env file
+    """
+    lines = []
+    if env.mtu:
+        lines.append(f"OCRD_PS_MTU={env.mtu}")
+    if env.ocrd_ps_port:
+        lines.append(f"OCRD_PS_PORT={env.ocrd_ps_port}")
+    if env.mongodb_user:
+        lines.append(f"MONGODB_USER={env.mongodb_user}")
+    if env.mongodb_pass:
+        lines.append(f"MONGODB_PASS={env.mongodb_pass}")
+    if env.mongodb_url:
+        lines.append(f"MONGODB_URL={env.mongodb_url}")
+    if env.rabbitmq_user:
+        lines.append(f"RABBITMQ_USER={env.rabbitmq_user}")
+    if env.rabbitmq_pass:
+        lines.append(f"RABBITMQ_PASS={env.rabbitmq_pass}")
+    if env.rabbitmq_url:
+        lines.append(f"RABBITMQ_URL={env.rabbitmq_url}")
+    if env.user_id:
+        lines.append(f"USER_ID={env.user_id}")
+    if env.group_id:
+        lines.append(f"GROUP_ID={env.group_id}")
+    if env.data_dir_host:
+        lines.append(f"DATA_DIR_HOST={env.data_dir_host}")
+    if env.internal_callback_url:
+        lines.append(f"INTERNAL_CALLBACK_URL={env.internal_callback_url}")
+    if env.run_network_dir:
+        lines.append(f"RUN_NETWORK_DIR={env.run_network_dir}")
+
+    with open(dest, "w+") as fout:
+        fout.write("\n".join(lines))
+
+
+def wait_for_startup(processing_server_url: str):
+    """Wait for completed startup of all docker-compose services
+
+    After the startup the containers need some time to be usable. This function ensures their
+    availability
+    """
+    counter = 0
+    while True:
+        try:
+            response = requests.get(processing_server_url)
+            response.raise_for_status()
+            break
+        except requests.exceptions.ConnectionError:
+            time.sleep(1)
+            counter += 1
+            if counter > 30:
+                raise Exception("processing-server startup failed") from None
+        except requests.HTTPError:
+            # unexpected error
+            exit(1)
+
+
+NETWORK_TEMPLATE = """
+networks:
+  default:
+    driver: bridge
+    driver_opts:
+      com.docker.network.driver.mtu: ${OCRD_PS_MTU}
+"""
+
+PROC_TEMPLATE = """
+  {service_name}:
+    image: {image}
+    container_name: {service_name}
+    command: {processor_name} worker --database $MONGODB_URL --queue $RABBITMQ_URL
+    depends_on: {depends_on}
+    user: "${{USER_ID}}:${{GROUP_ID}}"
+    profiles: [{profiles}]
+    volumes:
+      - "${{DATA_DIR_HOST}}:/data"
```

Comment:
@joschrew what about the processor resources, shouldn't that be a volume, too?

Preferably as **named volume** (so module-distributed and user-downloaded files can mix freely), e.g.:
```suggestion
    volumes:
      - "${{DATA_DIR_HOST}}:/data"
      - ocrd-resources:/usr/local/share/ocrd-resources
```

Unfortunately, it seems we missed the opportunity of the latest sweep across modules to define a unique internal resource location. In `ocrd/tesserocr`, we already use `/models` as alias for `/usr/local/share/ocrd-resources`, but in many other images we still use the latter...

Reply:
Perhaps brevity (`/models` instead of `/usr/local/share/ocrd-resources`) is not an issue if we are using scriped mounts and calls, anyway. And `/models` was a _hack_ of sorts: it was not backed by the spec (which still says /usr/local is the `system` location, and `$XDG_DATA_DIR` the `data` location), but only a convention for our fat container images.

Therefore I will switch ocrd/tesserocr back to /usr/local/share/ocrd-resources as the only place to mount (as in all the other _workable_ slim container images).

Reply:
> Therefore I will switch ocrd/tesserocr back to /usr/local/share/ocrd-resources as the only place to mount (as in all the other _workable_ slim container images).

https://github.com/OCR-D/ocrd_tesserocr/commit/70caca02260448cfbc7a76facc5bf6c4bd324691

Reply:
I have added a resource-volume now. What would you suggest to get the resources into the volume? I tried using the resmgr which fails because of missing writing permission on /.config (the setup cannot not use root permissions). Would it be a good idea to to this via the Makefile?

Reply:
> What would you suggest to get the resources into the volume? I tried using the resmgr which fails because of missing writing permission on /.config (the setup cannot not use root permissions). Would it be a good idea to to this via the Makefile?

Ah, that bit us before in the fat containers, so I thought we found a remedy for it: setting `XDG_CONFIG_HOME=/usr/local/share/ocrd-resources` as well, so the volume should cover both the resources themselves and the config file `ocrd/resources.yml`.

Your `/.config` indicates that `XDG_CONFIG_HOME` is not set (as `HOME=/`). What image is this  happening with?

So no, this should work on the user side already.

Reply:
I have now changed the ocrd-resources-volume mount to a lokal directory on the host instead of a named volume. This was to only way for me to use the resources with user's permissions which are necessary for correct permissions of workflow files

Reply:
All images. The setup is started as current user, this is necessary because the created workspace files should be of user's permission. If the volume is of root permission (or a named volume, same permissions) the resource download can only triggered by root. But when the container is started as user and then root user is used to download the resources we have logfile permission-errors again.

Reply:
Oh, right! We do have a problem with named volumes. They are always created as root. Even with `docker volume create`. It seems the only way to get normal ownership is by choosing a container-side parent path which already has the permissions we want (as these get inherited by the named volume). But in cases where the directory itself already exists (from the Docker build), its ownership won't change (i.e. stay root). That _is_ a big problem.

> The value of XDG_CONFIG_HOME has no effect on permissions.

It does have the effect of controlling where `ocrd/resources.yml` gets written.

Reply:
See https://github.com/moby/moby/issues/3124

Okay, so the only way out AFAICS is setting `/usr/local/share/ocrd-resources` (and its subdirectories) to 777 in the Dockerfile already. This means in _every_ Dockerfile.

You can test this: 
- revert https://github.com/OCR-D/ocrd_all/pull/449/commits/a9f2ffd65e36749980282e3d911659e3f58f94d3
- write a Dockerfile.fixup, contents:
```Dockerfile
ARG BASE_IMAGE
FROM $BASE_IMAGE
RUN mkdir -p /usr/local/share/ocrd-resources
RUN find /usr/local/share/ocrd-resources -type d -exec chmod 777 {} ";"
RUN find /usr/local/share/ocrd-resources -type f -exec chmod 666 {} ";"
```
- use it to patch a module of your choice, e.g.:
```sh
cat Dockerfile.fixup | docker build --build-arg BASE_IMAGE=ocrd/tesserocr -t ocrd/tesserocr:permissions -
```
- use the patched image, e.g.:
```sh
docker run --rm -it -v ocrd-resources:/usr/local/share/ocrd-resources ocrd/tesserocr:permissions ocrd resmgr download ocrd-tesserocr-recognize fra.traineddata
```

Reply:
I tried your suggestion, it nearly works for me (setting /usr/local/share/ocrd-resources itself to 777 is also necessary).
Another possibility would be to have more than one folder for the resources (kind of what we have currently). /usr/local/share/ocrd-resoucres could be used for preinstalled resources and the "on-demand"-resources could be installed to another folder (XDG_DATA_HOME/ocrd-resources for example) and for that folder a host-mounted-volume could be used. I'd prefer one folder for all resources but setting global permissions might rise security concerns again like it did when we talked about setting permissions for the logging-folders. Did we have a reason why we don't want to use XDG_DATA_HOME/ocrd-resources?

Reply:
> it nearly works for me (setting /usr/local/share/ocrd-resources itself to 777 is also necessary).

Are you sure you tried the exact command above (which should already cover the top-level directory as well)?

> Another possibility would be to have more than one folder for the resources [...]
> Did we have a reason why we don't want to use XDG_DATA_HOME/ocrd-resources?

Yes, we did have a reason. Not all processors can handle more than 1 location. Tesseract for example can not be brought to look up in more than 1 directory.

It's also not intuitive. You don't want to have to think about resources in terms of where they came from.